cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,929 @@
+
+      rmm::device_uvector<edge_t> num_triangles_invalid(num_invalid_edges, handle.get_stream());
+
+      // Update the number of triangles of each (p, q) edges by looking at their intersection
+      // size
+      thrust::adjacent_difference(handle.get_thrust_policy(),
+                                  intersection_offsets.begin() + 1,
+                                  intersection_offsets.end(),
+                                  num_triangles_invalid.begin());
+
+      thrust::for_each(
+        handle.get_thrust_policy(),
+        thrust::make_counting_iterator<edge_t>(0),
+        thrust::make_counting_iterator<edge_t>(num_invalid_edges),
+        unroll_edge_invalid<vertex_t, edge_t>{
+          edge_t{num_valid_edges},
+          raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+          raft::device_span<edge_t>(num_triangles_invalid.data(), num_triangles_invalid.size())});
+
+      size_t accumulate_pair_size = intersection_indices.size();
+
+      // FIXME: Find a way to not have to maintain a dataframe_buffer
+      auto vertex_pair_buffer_p_r_edge_p_q =
+        allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(accumulate_pair_size,
+                                                                     handle.get_stream());
+
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+        get_dataframe_buffer_end(vertex_pair_buffer_p_r_edge_p_q),
+        generate_p_r_q_r<vertex_t, edge_t, true>{
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          raft::device_span<vertex_t>(edgelist_srcs.data() + num_valid_edges, num_invalid_edges),
+          raft::device_span<vertex_t>(edgelist_dsts.data() + num_valid_edges, num_invalid_edges)});
+
+      auto num_edge_exists = accumulate_pair_size;
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(num_edge_exists),
+                       unroll_edge<vertex_t, edge_t, decltype(transposed_edge_first)>{
+                         edge_t{num_valid_edges},
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+                         transposed_edge_first,
+                         transposed_edge_first + num_valid_edges,
+                         transposed_edge_first + edgelist_srcs.size()});
+
+      auto vertex_pair_buffer_q_r_edge_p_q =
+        allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(accumulate_pair_size,
+                                                                     handle.get_stream());
```

Comment:
I think `vertex_pair_buffer_p_r_edge_p_q` is still necessary. In fact, nbr_intersection forced the invalid edges to be sorted by `src` and the derived intersection offsets and indices are necessary to populate  `vertex_pair_buffer_p_r_edge_p_q`. Same holds for `vertex_pair_buffer_q_r_edge_p_q` and in order to unroll both types of `vertex_pair_buffer`, I need the transposed edge list. instead of re-sorting back and forth, I prefer to sort it once before unrolling both `vertex_pair_buffer` types.